### PR TITLE
Preserve extra data in `RelationshipTarget` with `replace_related`(`_with_difference`)

### PR DESCRIPTION
**Pull request by urben1680**
# Objective

The methods and commands `replace_related` and `replace_related_with_difference` may cause data stored at the `RelationshipTarget` be lost when all original children are removed before new children are added.

Part of https://github.com/bevyengine/bevy/issues/19589

## Solution

Fix the issue, either by removing the old children _after_ adding the new ones and not _before_ (`replace_related_with_difference`) or by taking the whole `RelationshipTarget` to modify it, not only the inner collection (`replace_related`).

## Testing

I added a new test asserting the data is kept. I also added a general test of these methods as they had none previously.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Good; thanks for the tests here. This is too tricky to be confident without them.

File: crates/bevy_ecs/src/relationship/related_methods.rs [hunk 1]
```diff
@@ -668,4 +670,53 @@ mod tests {
...
+        let mut parent = world.spawn_empty();
+        parent.add_children(&[child1, child2]);
+
+        parent.replace_children(&[child2, child3]);
+        let children = parent.get::<Children>().unwrap().collection();
+        assert_eq!(children, &[child2, child3]);
+
+        parent.replace_children_with_difference(&[child3], &[child1, child2], &[child1]);
+        let children = parent.get::<Children>().unwrap().collection();
+        assert_eq!(children, &[child1, child2]);
```
Comment (urben1680) [inline crates/bevy_ecs/src/relationship/related_methods.rs]:
```suggestion
        let mut parent = world.spawn_empty();
        parent.add_children(&[child1, child2]);
        let some_child = Some(&ChildOf(parent.id()));

        parent.replace_children(&[child2, child3]);
        let children = parent.get::<Children>().unwrap().collection();
        assert_eq!(children, &[child2, child3]);
        assert_eq!(parent.world().get::<ChildOf>(child1), None);
        assert_eq!(parent.world().get::<ChildOf>(child2), some_child);
        assert_eq!(parent.world().get::<ChildOf>(child3), some_child);

        parent.replace_children_with_difference(&[child3], &[child1, child2], &[child1]);
        let children = parent.get::<Children>().unwrap().collection();
        assert_eq!(children, &[child1, child2]);
        assert_eq!(parent.world().get::<ChildOf>(child1), some_child);
        assert_eq!(parent.world().get::<ChildOf>(child2), some_child);
        assert_eq!(parent.world().get::<ChildOf>(child3), None);
```

File: crates/bevy_ecs/src/relationship/related_methods.rs [hunk 2]
```diff
@@ -668,4 +670,60 @@ mod tests {
...
+    #[test]
+    fn replace_related_works() {
+        let mut world = World::new();
+        let child1 = world.spawn_empty().id();
+        let child2 = world.spawn_empty().id();
+        let child3 = world.spawn_empty().id();
+
+        let mut parent = world.spawn_empty();
+        parent.add_children(&[child1, child2]);
+        let some_child = Some(&ChildOf(parent.id()));
```
Comment (urben1680) [inline crates/bevy_ecs/src/relationship/related_methods.rs]:
```suggestion
        let child_value = ChildOf(parent.id());
        let some_child = Some(&child_value);
```

File: crates/bevy_ecs/src/relationship/related_methods.rs [hunk 3]
```diff
@@ -245,20 +247,6 @@ impl<'w> EntityWorldMut<'w> {
...
-            for new_relation in newly_related_entities {
-                // We're changing the target collection manually so don't run the insert hook
-                world
-                    .entity_mut(*new_relation)
-                    .insert_with_relationship_hook_mode(R::from(this), RelationshipHookMode::Skip);
-            }
-        });
-
         if !entities_to_relate.is_empty() {
             if let Some(mut target) = self.get_mut::<R::RelationshipTarget>() {
```
Comment (gwafotapa) [inline crates/bevy_ecs/src/relationship/related_methods.rs]: Given the `if` at line 244, this `if let else` always branches on `if let`.

File: crates/bevy_ecs/src/relationship/related_methods.rs [hunk 4]
```diff
@@ -245,20 +247,6 @@ impl<'w> EntityWorldMut<'w> {
...
-
-            for new_relation in newly_related_entities {
-                // We're changing the target collection manually so don't run the insert hook
-                world
-                    .entity_mut(*new_relation)
-                    .insert_with_relationship_hook_mode(R::from(this), RelationshipHookMode::Skip);
-            }
-        });
-
         if !entities_to_relate.is_empty() {
```
Comment (gwafotapa) [inline crates/bevy_ecs/src/relationship/related_methods.rs]: I don't think this test is necessary.

**Review by gwafotapa — CHANGES_REQUESTED** [review]
Looks good to me! I would just like to see the dead code disappear.

On a side note, I'm not sure why `replace_related_with_difference` needs the `entities_to_relate` argument.

**Comment by urben1680** [comment]
> On a side note, I'm not sure why replace_related_with_difference needs the entities_to_relate argument.

It is more efficient to reinsert them after clearing the `RelationshipTarget`. Otherwise you would have to call `RelationshipSourceCollection::remove` in a loop which is quite inefficient for `Vec<Entity>`. It is a better idea to just insert them again, also because the typical `Vec<Entity>` can be easily extended via a slice.

Also you need it in case the `RelationshipTarget` is missing despite the caller's assumption. If the first reason was not present, this would have to panic in that case.

Good call on the branches. :+1:

**Comment by gwafotapa** [comment]
Ah I see. Thank you.

**Review by gwafotapa — APPROVED** [review]
